pf/internal/defaults/defaults.go

```diff
@@ -0,0 +1,293 @@
+	TopSchemaMap shim.SchemaMap
+
+	// Toplevel SchemaInfo configuration matching TopSchemaMap.
+	TopFieldInfos map[string]*tfbridge.SchemaInfo
+
+	// Optional. Note that ResourceInstance need not be specified for Invoke or Configure processing.
+	ResourceInstance *tfbridge.PulumiResource
+
+	// Optional. If known, these are the provider-level configuration values, to support DefaultInfo.Config.
+	ProviderConfig resource.PropertyMap
+}
+
+// Transforms a PropertyMap to apply default values specified in DefaultInfo.
+//
+// These values are specified at the bridged provider configuration level, and are applied before any Terraform
+// processing; therefore the function works at Pulumi level (transforming a PropertyMap).
+func ApplyDefaultInfoValues(ctx context.Context, args ApplyDefaultInfoValuesArgs) resource.PropertyMap {
+
+	// Can short-circuit the entire processing if there are no matching SchemaInfo entries, and therefore no
+	// matching DefaultInfo entries at all.
+	if args.TopFieldInfos == nil {
+		return args.PropertyMap
+	}
+
+	t := &defaultsTransform{
+		resourceInstance: args.ResourceInstance,
+		topSchemaMap:     args.TopSchemaMap,
+		topFieldInfos:    args.TopFieldInfos,
+		providerConfig:   args.ProviderConfig,
+	}
+	result := t.withDefaults(ctx, make(resource.PropertyPath, 0), resource.NewObjectProperty(args.PropertyMap))
+	if !result.IsObject() {
+		contract.Failf("defaultsTransform.withDefaults returned a non-object value")
+	}
+	return result.ObjectValue()
+}
+
+func getDefaultValue(
+	ctx context.Context,
+	res *tfbridge.PulumiResource,
+	fieldSchema shim.Schema,
+	defaultInfo *tfbridge.DefaultInfo,
+	providerConfig resource.PropertyMap,
+) (resource.PropertyValue, bool, error) {
+	na := resource.NewNullProperty()
+
+	if defaultInfo == nil {
+		return na, false, nil
+	}
+	if defaultInfo.From != nil && res != nil {
+		raw, err := defaultInfo.From(res)
+		if err != nil {
+			return na, false, err
+		}
+		if raw == nil {
+			return na, false, nil
+		}
+		return recoverDefaultValue(raw), true, nil
+
+	} else if defaultInfo.EnvVars != nil {
+		for _, n := range defaultInfo.EnvVars {
+			if str, ok := os.LookupEnv(n); ok {
+				// TODO what is the behavior of an env var set to an empty string? Is this the same as
+				// unset, or the same as actually setting to empty?
+				v, err := parseValueFromEnv(fieldSchema, str)
+
+				tflog.Info(ctx,
+					"Applying a default value from an environment variable",
+					map[string]any{
+						"envvar": n,
+					})
+
+				return v, true, err
+			}
+		}
+	} else if defaultInfo.Config != "" {
+		pk := resource.PropertyKey(defaultInfo.Config)
+		if providerConfig != nil {
+			if pv, ok := providerConfig[pk]; ok {
+				return pv, true, nil
+			}
+		}
+	}
+
+	if defaultInfo.Value != nil {
+		return recoverDefaultValue(defaultInfo.Value), true, nil
+	}
+	return na, false, nil
+}
+
+func parseValueFromEnv(sch shim.Schema, str string) (resource.PropertyValue, error) {
```

Comment:
This interprets `parseValueFromEnv(shim.TypeNumber, "")` as `0.0`. It feels like this should be an error or unset, but definitely not a zero value.

Reply:
You're right; I've looked at prod code and it dismisses env vars that are set by empty; copying this behavior for now. Can add a quick test.

Reply:
Since we are no longer tolerating empty `str` values, let's (1) remove the `if str != ""` checks and (2) add a `contract.Assertf(str != "", "empty env vars should have been skipped")` at the top of the function. This makes the intended behavior of the function much easer to read at a glance.